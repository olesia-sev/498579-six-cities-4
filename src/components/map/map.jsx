import React, {useEffect, useMemo, useRef} from "react";
import PropTypes from 'prop-types';
import leaflet from 'leaflet';
import {connect} from "react-redux";
import {offersTypeArray, offerType} from "../../prop-types/prop-types";

const mapConfig = {
  center: [52.38333, 4.9],
  zoom: 12,
  zoomControl: false,
  marker: true,
  city: [52.38333, 4.9],
  mapIcon: {
    iconUrl: `/img/pin.svg`,
    iconSize: [30, 30]
  },
  mapActiveIcon: {
    iconUrl: `/img/pin-active.svg`,
    iconSize: [30, 30]
  }
};

const icon = leaflet.icon(mapConfig.mapIcon);
const iconActive = leaflet.icon(mapConfig.mapActiveIcon);

const renderMap = () => {
  const map = leaflet.map(`map`, mapConfig);

  map.setView(mapConfig.city, mapConfig.zoom);

  leaflet
    .tileLayer(`https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`, {
      attribution: `&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>`
    })
    .addTo(map);

  return map;
};

const Map = React.memo(function Map({offers, hoveredOffer, activeCityId}) {
  const mapRef = useRef(null);

  const currentCityOffers = useMemo(() => offers.filter(
      (offer) => offer.cityId === activeCityId),
  [activeCityId, offers]
  );

  useEffect(() => {
    mapRef.current = renderMap();
    return () => {
      mapRef.current.remove();
    };
  }, []);

  useEffect(() => {
    const coords = currentCityOffers.map((offer) => offer.coords);

    const placemarks = [];

    for (let i = 0; i < coords.length; i++) {
      const isActive = hoveredOffer && coords[i] === hoveredOffer.coords;
      const placemark = leaflet.marker(coords[i], {icon: isActive ? iconActive : icon})
        .addTo(mapRef.current);

      placemarks.push(placemark);
    }

    return () => {
      placemarks.forEach((placemark) => {
        mapRef.current.removeLayer(placemark);
      });
    };
  }, [currentCityOffers, hoveredOffer]);

  return (
    <div id="map" />
  );
});

Map.propTypes = {
  offers: offersTypeArray,
  hoveredOffer: offerType,
  activeCityId: PropTypes.number.isRequired,
};

const mapStateToProps = ({hoveredOffer, offers, activeCityId}) => {
  return {hoveredOffer, offers, activeCityId};
};

export default connect(mapStateToProps)(Map);
