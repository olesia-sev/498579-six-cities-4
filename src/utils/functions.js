export const getRandomNumberInRange = (min, max) => {
  return Math.floor(Math.random() * (max - min + 1)) + min;
};

export const extend = (a, b) => {
  return Object.assign({}, a, b);
};
