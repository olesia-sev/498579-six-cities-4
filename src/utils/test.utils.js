export const optionsAmount = 100;

export const testString = `Test string`;

export const emptyFunction = () => {};

export const cardDataArray = [
  {
    id: 10,
    img: `https://placedog.net/260/200?id=2`,
    images: [
      `https://placedog.net/260/200?id=3`,
      `https://placedog.net/260/200?id=33`,
      `https://placedog.net/260/200?id=12`,
      `https://placedog.net/260/200?id=10`,
      `https://placedog.net/260/200?id=11`,
      `https://placedog.net/260/200?id=15`,
    ],
    price: 150,
    title: `Title`,
    placeType: `Flat`,
    numberOfBedrooms: 2,
    maxGuests: 3,
    rating: 4,
    saved: true,
    premium: true,
    featuresInside: [
      `Wi-Fi`, `Heating`,
    ],
    hostName: `Pete`,
    hostAvatar: `https://placedog.net/260/200?id=5`,
    userPro: true,
    description: `Description`,
  },
  {
    id: 100,
    img: `https://placedog.net/260/200?id=3`,
    images: [
      `https://placedog.net/260/200?id=33`,
      `https://placedog.net/260/200?id=99`,
      `https://placedog.net/260/200?id=89`,
      `https://placedog.net/260/200?id=29`,
      `https://placedog.net/260/200?id=13`,
      `https://placedog.net/260/200?id=16`,
    ],
    price: 150,
    title: `Title test`,
    placeType: `Flat`,
    numberOfBedrooms: 1,
    maxGuests: 2,
    rating: 4,
    saved: false,
    premium: false,
    featuresInside: [
      `Wi-Fi`, `Heating`,
    ],
    hostName: `Pete`,
    hostAvatar: `https://placedog.net/260/200?id=57`,
    userPro: false,
    description: `Description`,
  }
];
