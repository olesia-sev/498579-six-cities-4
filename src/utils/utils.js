export const extend = (a, b) => {
  return Object.assign({}, a, b);
};

export const SORT_TYPES = {
  POPULAR: `Popular`,
  TO_HIGH_PRICE: `Price: low to high`,
  TO_LOW_PRICE: `Price: high to low`,
  TOP_RATED: `Top rated first`,
};

export const getSortedOffers = (offers, activeSort) => {
  switch (activeSort) {
    case SORT_TYPES.POPULAR:
      return offers.slice();
    case SORT_TYPES.TO_HIGH_PRICE:
      return offers.slice().sort((a, b) => a.price - b.price);
    case SORT_TYPES.TO_LOW_PRICE:
      return offers.slice().sort((a, b) => b.price - a.price);
    case SORT_TYPES.TOP_RATED:
      return offers.slice().sort((a, b) => b.rating - a.rating);
    default:
      return offers.slice();
  }
};

export const MAX_REVIEWS_LENGTH = 10;

export const AppRoute = {
  LOGIN: `/login`,
  ROOT: `/`,
  FAVOURITES: `/favorites`,
  OFFER_DETAIL: `/offers/:id`,
};

export const RATING_STAR_WIDTH = 20;

export const MAX_IMAGES_AMOUNT = 6;
export const SLICE_BEGIN = 0;
