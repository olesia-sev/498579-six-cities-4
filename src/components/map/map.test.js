import React from "react";
import renderer from "react-test-renderer";
import {cardDataArray} from "../../utils/test.utils";
import {BrowserRouter as Router} from "react-router-dom";
import {Map} from "./map";

it(`Map container should be rendered`, () => {
  const tree = renderer
    .create(<Router>
      <Map
        offers = {cardDataArray}
      />
    </Router>)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
