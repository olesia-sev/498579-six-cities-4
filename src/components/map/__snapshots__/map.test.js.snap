// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map container should be rendered 1`] = `
<div
  id="map"
/>
`;
